jsonschema/validators.py

```diff
@@ -145,16 +145,23 @@ def schema_validators(self, schema):
             """
             Returns a list of validators that should apply for the given schema
             """
-            legacy_cls = [Draft3Validator, Draft4Validator, Draft6Validator, Draft7Validator]
+            legacy_validator_cls = [
+                Draft3Validator,
+                Draft4Validator,
+                Draft6Validator,
+                Draft7Validator,
+            ]
 
             ref = schema.get(u"$ref")
             if ref is not None:
                 # We make sure $ref is evaluated first if available in schema
                 validators = [(u"$ref", ref)]
 
                 # Apply all remaining validators for newer drafts
-                if not any(isinstance(self, x) for x in legacy_cls):
-                    validators += [(x, schema[x]) for x in schema if x not in ["$ref"]]
+                if not any(isinstance(self, x) for x in legacy_validator_cls):
```

Comment:
This still doesn't work for me as I have a class extended from Draft7Validator. But Draft7Validator is not a type. It's type is 'Validator', right? So does this ever work?

The relevant code for me is here:
https://github.com/devicetree-org/dt-schema/blob/master/dtschema/lib.py#L668

Reply:
The existing test cases don't fail with this construct, but I will check this deeper. Maybe you have a sugestion how to handle this better :slightly_smiling_face:

Reply:
I think you need a parameter for create() to handle the different behavior. Then any validator created with extend() inherits the same behavior if not specified.

Reply:
Moved this into two functions which can now be used when creating a validator. What do you think about this approach?

@Julian can `_schema_validators_draft7_draft6_draft_4_draft3` be moved to `_legacy_validators.py` or would it maked sence to create a `_legacy` module?

Reply:
If I follow, the code's changed, in the current version what you have looks good to me @nezhar, but let's call it `applicable_validators`, and then yes you can move the one that ignores non-`$ref` to `_legacy_validators`.

While I see it though, @robherring just FYI, subclassing validators isn't part of their public API (and isn't supported), so there's no promises stuff continues to work there. If it works now though I guess great.

Reply:
Renamed to `applicable_validators` and moved to old code to `_legacy_validators`